SimPEG/VRM/ProblemVRM.py

```diff
@@ -155,170 +140,436 @@ def _getGeometryMatrix(self, xyzc, xyzh, pp):
-                for rr in range(0, M):
-                    u1 = locs[rr, 0] - ax
-                    u1[np.abs(u1) < tol] = np.min(xyzh[:, 0])/tol2
-                    u2 = locs[rr, 0] - bx
-                    u2[np.abs(u2) < tol] = -np.min(xyzh[:, 0])/tol2
-                    v1 = locs[rr, 1] - ay
-                    v1[np.abs(v1) < tol] = np.min(xyzh[:, 1])/tol2
-                    v2 = locs[rr, 1] - by
-                    v2[np.abs(v2) < tol] = -np.min(xyzh[:, 1])/tol2
-                    w1 = locs[rr, 2] - az
-                    w1[np.abs(w1) < tol] = np.min(xyzh[:, 2])/tol2
-                    w2 = locs[rr, 2] - bz
-                    w2[np.abs(w2) < tol] = -np.min(xyzh[:, 2])/tol2
-
-                    Gxx = (
-                        np.arctan((v1*w1)/(u1*np.sqrt(u1**2+v1**2+w1**2)+tol)) -
-                        np.arctan((v1*w1)/(u2*np.sqrt(u2**2+v1**2+w1**2)+tol)) +
-                        np.arctan((v2*w1)/(u2*np.sqrt(u2**2+v2**2+w1**2)+tol)) -
-                        np.arctan((v2*w1)/(u1*np.sqrt(u1**2+v2**2+w1**2)+tol)) +
-                        np.arctan((v2*w2)/(u1*np.sqrt(u1**2+v2**2+w2**2)+tol)) -
-                        np.arctan((v1*w2)/(u1*np.sqrt(u1**2+v1**2+w2**2)+tol)) +
-                        np.arctan((v1*w2)/(u2*np.sqrt(u2**2+v1**2+w2**2)+tol)) -
-                        np.arctan((v2*w2)/(u2*np.sqrt(u2**2+v2**2+w2**2)+tol))
+            if isinstance(rxObj, Point):
+
+                if dComp.lower() == 'x':
+                    for rr in range(0, M):
+                        u1 = locs[rr, 0] - ax
+                        u1[np.abs(u1) < tol] = np.min(xyzh[:, 0])/tol2
+                        u2 = locs[rr, 0] - bx
+                        u2[np.abs(u2) < tol] = -np.min(xyzh[:, 0])/tol2
+                        v1 = locs[rr, 1] - ay
+                        v1[np.abs(v1) < tol] = np.min(xyzh[:, 1])/tol2
+                        v2 = locs[rr, 1] - by
+                        v2[np.abs(v2) < tol] = -np.min(xyzh[:, 1])/tol2
+                        w1 = locs[rr, 2] - az
+                        w1[np.abs(w1) < tol] = np.min(xyzh[:, 2])/tol2
+                        w2 = locs[rr, 2] - bz
+                        w2[np.abs(w2) < tol] = -np.min(xyzh[:, 2])/tol2
+
+                        Gxx = (
+                            np.arctan((v1*w1)/(u1*np.sqrt(u1**2+v1**2+w1**2)+tol)) -
+                            np.arctan((v1*w1)/(u2*np.sqrt(u2**2+v1**2+w1**2)+tol)) +
+                            np.arctan((v2*w1)/(u2*np.sqrt(u2**2+v2**2+w1**2)+tol)) -
+                            np.arctan((v2*w1)/(u1*np.sqrt(u1**2+v2**2+w1**2)+tol)) +
+                            np.arctan((v2*w2)/(u1*np.sqrt(u1**2+v2**2+w2**2)+tol)) -
+                            np.arctan((v1*w2)/(u1*np.sqrt(u1**2+v1**2+w2**2)+tol)) +
+                            np.arctan((v1*w2)/(u2*np.sqrt(u2**2+v1**2+w2**2)+tol)) -
+                            np.arctan((v2*w2)/(u2*np.sqrt(u2**2+v2**2+w2**2)+tol))
+                        )
```

Comment:
The first thing that jumped into my eye is that there are a LOT of repeated operations. (This is not new with your pull request, I saw that it was that way before. However, this PR would be a chance to clean it up.) Now, to be honest, I have no idea if these functions are time critical or not. But by simply re-arranging things I could get a speed-up of at least 40 % in the creation of the G-vectors. Have a look here: https://github.com/prisae/tmp-share/blob/master/timing-test_combine-common-operations.ipynb.

Specifically, I am talking about all the repeated `u1**2` etc. Making once `su1 = u1*u1` and use `su1` later on will save time (`u1*u1` is a tad faster than `u1**2`, but that doesn't add up. But `su1=u1**2; su1; su1; su1` is significantly faster than `u1**2; u1**2; u1**2`). The same applies for the repeated `np.sqrt(...)`.

As these could even be placed outside the `if dComp.lower() == 'x'/'y'/'z'`, it would also save a lot writing overall.

As I said, just an idea/suggestion.